README.md
## Scaffolding

- Install GT `sudo npm i -g granturismo`

- `gt config add vivaxy/gt-front-end-scaffold https://github.com/vivaxy/gt-front-end-scaffold.git`

- `gt init`

- Select `vivaxy/gt-front-end-scaffold`

----------
__________

## Initialization

`yarn install`

## Development

`yarn run dev`

## Build

`yarn run build`

basic configuration files for front end projects

```
├── dist              # 可执行代码目录，需要加入 .gitignore
│   ├── css           # css 输出文件
│   ├── html          # html 输出文件
│   └── js            # js 输出文件
├── node_modules      # node 依赖，需要加入 .gitignore
├── scripts           # 工具
│   ├── config.js     # wepback 配置
│   └── dev.js        # webpack 本地调试服务
├── src               # 源码目录
│   ├── ejs           # 模版文件
│   ├── entries       # js 入口文件，每个入口都会生成一个对应的 html
│   ├── lib           # js 基础库
│   ├── styles        # 样式源文件 postcss
│   └── templates     # html 模版文件
├── .babelrc          # babel 配置
├── .editorconfig     # editorconfig 配置文件
├── .gitignore        # git 忽略文件
├── CHANGELOG.md      # 改动日志
├── LICENSE           # 开源许可证
├── package.json      # node 配置文件
├── README.md         # 说明文档
├── webpack.config.js # webpack 配置文件
└── yarn.lock         # yarn 版本锁
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.2.1"></a>
## [0.2.1](https://github.com/vivaxy/gt-front-end-scaffold/compare/v0.2.0...v0.2.1) (2018-01-30)


### Bug Fixes

* **gt:** :bug:Support prettier within gt scripts ([1020754](https://github.com/vivaxy/gt-front-end-scaffold/commit/1020754))
* **prettier:** :bug:Fix format package.json ([87244bb](https://github.com/vivaxy/gt-front-end-scaffold/commit/87244bb))



<a name="0.2.0"></a>
# [0.2.0](https://github.com/vivaxy/gt-front-end-scaffold/compare/v0.1.1...v0.2.0) (2018-01-26)


### Features

* **ejs:** :sparkles:Remove ejs template feature ([d14ef05](https://github.com/vivaxy/gt-front-end-scaffold/commit/d14ef05))
* **prettier:** :sparkles:Add prettier formatter ([0807393](https://github.com/vivaxy/gt-front-end-scaffold/commit/0807393))



<a name="0.1.1"></a>

## [0.1.1](https://github.com/vivaxy/gt-front-end-scaffold/compare/v0.1.0...v0.1.1) (2018-01-11)

<a name="0.1.0"></a>

# 0.1.0 (2017-03-17)

### Bug Fixes

* **gt:** :bug:remove es6 in gt.js ([7dd98a3](https://github.com/vivaxy/gt-front-end-scaffold/commit/7dd98a3))
* **gt:** fix es6 in gt ([3421c0a](https://github.com/vivaxy/gt-front-end-scaffold/commit/3421c0a))
* **webpack:** :bug:Rename templates folder, Remove eslint, Remove tests, Update README.md ([25f7d71](https://github.com/vivaxy/gt-front-end-scaffold/commit/25f7d71))

### Features

* **gt:** :sparkles:Upgrade gt.js ([f27802a](https://github.com/vivaxy/gt-front-end-scaffold/commit/f27802a))
* **webpack:** :sparkles:Add webpack ([d45f405](https://github.com/vivaxy/gt-front-end-scaffold/commit/d45f405))

# 0.0.0

initialize project with everything i got


test/index.js
/**
 * @since 2015-12-27 21:13
 * @author vivaxy
 */

console.log('ok'); // eslint-disable-line no-console, strict


scripts/dev.js
/**
 * @since 2017-02-20 11:58
 * @author vivaxy
 */

const open = require('open');
const webpack = require('webpack');
const WebpackDevServer = require('webpack-dev-server');

const webpackConfig = require('../webpack.config');
const config = require('./config');

const compiler = webpack(webpackConfig);

let opened = false;
const entry = 'index';

const devServerOptions = {
    hot: true,
    historyApiFallback: true,
    stats: {
        colors: true
    },
    contentBase: config.DIST_PATH
};

const openBrowser = () => {
    const address = server.listeningApp.address();
    const url = `http://${address.address}:${address.port}`;
    console.log(`   server started: ${url}`);
    open(`${url}/html/${entry}.html`);
};

const server = new WebpackDevServer(compiler, devServerOptions);

compiler.plugin('done', function() {
    if (!opened) {
        opened = true;
        openBrowser();
    }
});

server.listen(config.DEVELOPMENT_PORT, config.DEVELOPMENT_IP, function(err) {
    if (err) {
        console.log(err);
    }
});

const stdIn = process.stdin;
stdIn.setEncoding('utf8');
stdIn.on('data', openBrowser);


webpack.config.js
'use strict';

const path = require('path');
const glob = require('glob');
const numeral = require('numeral');
const webpack = require('webpack');
const logUpdate = require('log-update');
const autoprefixer = require('autoprefixer');
const Visualizer = require('webpack-visualizer-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');

const config = require('./scripts/config');

const SOURCE_PATH = 'src';
const ENTRY_FOLDER_NAME = 'entries';
const TEMPLATE_FOLDER_NAME = 'templates';
const COMMON_CHUNK_NAME = 'common';
const ENV_DEVELOPMENT = 'development';
const ENV_PRODUCTION = 'production';
const BANNER = '@2017 vivaxy';

let NODE_ENV = process.env.NODE_ENV || ENV_PRODUCTION;

const postcssLoader = {
    loader: 'postcss-loader',
    options: {
        plugins: function() {
            return [autoprefixer];
        }
    }
};

const urlLoader = {
    loader: 'url-loader',
    options: {
        limit: 8192,
        name: 'image/[name]-[hash].[ext]'
    }
};

const jsRule = {
    test: /\.js$/,
    use: ['babel-loader']
};

const pcssRule = {
    test: /\.pcss$/,
    include: [path.resolve(__dirname, SOURCE_PATH)],
    use: ['style-loader', 'css-loader', postcssLoader]
};

const imageRule = {
    test: /\.(png|jpg|gif)$/,
    use: [urlLoader]
};

const ejsRule = {
    test: /\.ejs$/,
    use: ['raw-loader']
};

const webpackConfig = {
    entry: {
        [COMMON_CHUNK_NAME]: ['babel-polyfill']
    },
    output: {
        path: path.resolve(__dirname, config.DIST_PATH),
        filename: 'js/[name].js',
        publicPath: '../'
    },
    module: {
        rules: [jsRule, pcssRule, imageRule, ejsRule]
    },
    plugins: [
        new webpack.EnvironmentPlugin(['NODE_ENV']),
        new webpack.optimize.CommonsChunkPlugin({
            name: COMMON_CHUNK_NAME,
            filename: 'js/[name].js',
            minChunks: 2 // Infinity
        }),
        new Visualizer(),
        new webpack.NamedModulesPlugin()
    ]
};

const entryFileNameList = glob.sync(
    path.join(SOURCE_PATH, ENTRY_FOLDER_NAME) + '/*.js'
);
const entryNameList = entryFileNameList.map(function(entryFileName) {
    return path.basename(entryFileName, '.js');
});

// get corresponding html template
const htmlFileNameList = glob.sync(
    path.join(SOURCE_PATH, TEMPLATE_FOLDER_NAME) + '/*.html'
);
const htmlNameList = htmlFileNameList.map(function(htmlFileName) {
    return path.basename(htmlFileName, '.html');
});

// set entry
entryNameList.forEach(function(entryName) {
    webpackConfig.entry[entryName] = [
        path.join(__dirname, SOURCE_PATH, ENTRY_FOLDER_NAME, entryName + '.js')
    ];

    let htmlTemplateName = 'index';
    if (htmlNameList.indexOf(entryName) !== -1) {
        htmlTemplateName = entryName;
    }

    webpackConfig.plugins.push(
        new HtmlWebpackPlugin({
            template: `${SOURCE_PATH}/${TEMPLATE_FOLDER_NAME}/${htmlTemplateName}.html`,
            filename: `html/${entryName}.html`,
            hash: false,
            inject: 'body',
            chunks: [COMMON_CHUNK_NAME, entryName]
        })
    );
});

switch (NODE_ENV) {
    case ENV_DEVELOPMENT:
        entryNameList.forEach(function(entryName) {
            webpackConfig.entry[entryName].unshift(
                `webpack-dev-server/client?http://${config.DEVELOPMENT_IP}:${
                    config.DEVELOPMENT_PORT
                }`
            );
            webpackConfig.entry[entryName].unshift('webpack/hot/dev-server');
        });

        webpackConfig.devtool = 'eval';

        webpackConfig.plugins.push(new webpack.HotModuleReplacementPlugin());
        webpackConfig.plugins.push(new webpack.NoEmitOnErrorsPlugin());
        webpackConfig.plugins.push(
            new webpack.ProgressPlugin((percentage, msg) => {
                logUpdate(
                    '     progress:',
                    numeral(percentage).format('00.00%'),
                    msg
                );
            })
        );

        break;

    case ENV_PRODUCTION:
        webpackConfig.devtool = `source-map`;

        webpackConfig.plugins.push(
            new webpack.optimize.UglifyJsPlugin({
                sourceMap: true
            })
        );
        webpackConfig.plugins.push(
            new webpack.BannerPlugin({
                banner: BANNER
            })
        );
        break;
}

module.exports = webpackConfig;


scripts/gt/index.js
/**
 * @since 2017-03-21 15:59:47
 * @author vivaxy
 */

const Listr = require('listr');

const sleep = timeout => {
    return new Promise(resolve => {
        setTimeout(resolve, timeout);
    });
};

const copyFiles = options => {
    const { presets } = options;

    const files = [
        'scripts/config.js',
        'scripts/dev.js',
        'src',
        '.babelrc',
        '.editorconfig',
        '.gitignore',
        '.prettierignore',
        'CHANGELOG.md',
        'LICENSE',
        'webpack.config.js',
        'yarn.lock'
    ];

    return async () => {
        await sleep(1000);
        await presets.copyFiles(files);
    };
};

const updatePackageJSON = options => {
    const { project, presets } = options;
    const projectGit = project.git || {};
    const filename = 'package.json';

    return async () => {
        await sleep(1000);
        await presets.updateJson(filename, data => {
            const {
                // to remove
                version,
                repository,

                // not changed
                ...rest
            } = data;

            return {
                ...rest,
                name: project.name,
                version: '0.0.0',
                reactScaffoldVersion: version,
                repository: {
                    type: repository.type,
                    url: projectGit.repositoryURL
                },
                bugs: {
                    url: projectGit.repositoryURL
                },
                homepage: projectGit.repositoryURL
            };
        });
    };
};

const updateREADME = options => {
    const { project, presets } = options;
    const filename = 'README.md';

    return async () => {
        await sleep(1000);
        await presets.updateFile(filename, data => {
            const projectData = data.split('----------')[1];
            return projectData.replace(
                /__________/g,
                `${project.name}

Initialized by [vivaxy/gt-front-end-scaffold](https://github.com/vivaxy/gt-front-end-scaffold)`
            );
        });
    };
};

const updateCHANGELOG = options => {
    const { presets } = options;
    const filename = 'CHANGELOG.md';

    return async () => {
        await sleep(1000);
        await presets.updateFile(filename, data => {
            return '';
        });
    };
};

export const init = options => {
    return new Listr([
        {
            title: 'copy files',
            task: copyFiles(options)
        },
        {
            title: 'update package.json',
            task: updatePackageJSON(options)
        },
        {
            title: 'update README.md',
            task: updateREADME(options)
        },
        {
            title: 'update CHANGELOG.md',
            task: updateCHANGELOG(options)
        }
    ]);
};

export const after = () => {
    console.log(`
        - Run 'yarn install' to initialize the project.
        - Run 'yarn dev' to start a development server.
        - Run 'yarn build' to build release files.
`);
};
